lib/auth/export.go

```diff
@@ -0,0 +1,228 @@
+	case "db":
+		return exportTLSAuthority(ctx, req.Client, types.DatabaseCA, false, req.ExportPrivateKeys)
+	case "tls-user-der", "windows":
+		return exportTLSAuthority(ctx, req.Client, types.UserCA, true, req.ExportPrivateKeys)
+	}
+
+	// if no --type flag is given, export HostCA and UserCA.
+	if req.AuthType == "" {
+		typesToExport = []types.CertAuthType{types.HostCA, types.UserCA}
+	} else {
+		authType := types.CertAuthType(req.AuthType)
+		if err := authType.Check(); err != nil {
+			return "", trace.Wrap(err)
+		}
+		typesToExport = []types.CertAuthType{authType}
+	}
+	localAuthName, err := req.Client.GetDomainName(ctx)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	// fetch authorities via auth API (and only take local CAs, ignoring
+	// trusted ones)
+	var authorities []types.CertAuthority
+	for _, at := range typesToExport {
+		cas, err := req.Client.GetCertAuthorities(ctx, at, req.ExportPrivateKeys)
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+		for _, ca := range cas {
+			if ca.GetClusterName() == localAuthName {
+				authorities = append(authorities, ca)
+			}
+		}
+	}
+
+	for _, ca := range authorities {
+		if req.ExportPrivateKeys {
+			for _, key := range ca.GetActiveKeys().SSH {
+				fingerprint, err := sshutils.PrivateKeyFingerprint(key.PrivateKey)
+				if err != nil {
+					return "", trace.Wrap(err)
+				}
+				if req.ExportAuthorityFingerprint != "" && fingerprint != req.ExportAuthorityFingerprint {
+					continue
+				}
+				ret.WriteString(string(key.PrivateKey) + "\n")
+			}
+		} else {
+			for _, key := range ca.GetTrustedSSHKeyPairs() {
+				fingerprint, err := sshutils.AuthorizedKeyFingerprint(key.PublicKey)
+				if err != nil {
+					return "", trace.Wrap(err)
+				}
+				if req.ExportAuthorityFingerprint != "" && fingerprint != req.ExportAuthorityFingerprint {
+					continue
+				}
+
+				// export certificates in the old 1.0 format where host and user
+				// certificate authorities were exported in the known_hosts format.
+				if req.CompatVersion == "1.0" {
+					castr, err := hostCAFormat(ca, key.PublicKey, req.Client)
+					if err != nil {
+						return "", trace.Wrap(err)
+					}
+
+					ret.WriteString(castr + "\n")
+					continue
+				}
+
+				// export certificate authority in user or host ca format
+				var castr string
+				switch ca.GetType() {
+				case types.UserCA:
+					castr, err = userCAFormat(ca, key.PublicKey)
+				case types.HostCA:
+					castr, err = hostCAFormat(ca, key.PublicKey, req.Client)
+				default:
+					return "", trace.BadParameter("unknown user type: %q", ca.GetType())
+				}
+				if err != nil {
+					return "", trace.Wrap(err)
+				}
+
+				// write the export friendly string
+				ret.WriteString(castr + "\n")
+			}
+		}
+	}
+
+	return ret.String(), nil
+}
+
+func exportTLSAuthority(ctx context.Context, client ClientI, typ types.CertAuthType, unpackPEM bool, exportPrivateKeys bool) (string, error) {
```

Comment:
nit1:

```suggestion
func exportTLSAuthority(ctx context.Context, client ClientI, authType types.CertAuthType, unpackPEM bool, exportPrivateKeys bool) (string, error) {
```

Suggestion/nit: 5 params is getting into "too many params" territory. Consider grouping authType, unpackPEM and exportPrivateKeys in a single struct - this also has a side benefit that we can use field labels to document the booleans in call sites.